### DISCUSSION TITLE
Lint incompatible module loading mechanism (CJS vs ESM)

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#500] [RFC]
You are doomed if you see `ERR_REQUIRE_ESM` or similar error messages with `ESM`, you'll spend hours googling and trying out different solutions to make things work.

As a linter, I think we may be able help with the situation by figuring out your incompatible usage and tell you what to do?

Similar project: https://github.com/arethetypeswrong/arethetypeswrong.github.io

My first questions are:
* Is this a good idea?
* Is this doable (as a linter written in Rust)?

**Comment by Boshen** [comment]
Response from Andrew: https://twitter.com/atcb/status/1674299679063015424

**Comment by strager** [comment]
> Is this a good idea?

Yes. I don't know if it's a good idea for Oxc, though.

> Is this doable (as a linter written in Rust)?

Why wouldn't it be doable?

**Comment by andrewbranch** [comment]
I have a related idea that I think would be super useful. If you land functionality like this, and if you build a full module graph, it seems like you would be able to help authors detect and avoid the dual-package hazard. For example, if a project has a tree like

```
my-esm-project/
  node_modules/
    cjs-only-dep
    dual-dep
```

and a file like

```js
// index.mjs
import {} from "cjs-only-dep";
import {} from "dual-dep";
```

Now suppose that `cjs-only-dep` also has a dependency on `dual-dep`. `cjs-only-dep` likely accesses the CJS format of `dual-dep` via `require`, while `index.mjs` accesses the ESM format of `dual-dep`. So the user has included two different formats of the same package. But this tends not to be easily visible to the user until they hit mysterious problems.
